### ISSUE TITLE
wasm-sourcemap.py fails on Windows if Emscripten is installed to some other drive than C:\

### ISSUE THREAD

**Issue by juj** [emscripten-core/emscripten#10154]
Running e.g. `python tests\runner.py browser.test_pthread_lsan_leak` with Emscripten residing on drive E:\ rather than C:\ gives
```
test_pthread_lsan_leak (test_browser.browser) ...
Traceback (most recent call last):
  File "E:\code\emsdk\emscripten\incoming\tools\wasm-sourcemap.py", line 355, in <module>
    sys.exit(main())
  File "E:\code\emsdk\emscripten\incoming\tools\wasm-sourcemap.py", line 334, in main
    map = build_sourcemap(entries, code_section_offset, prefixes, options.sources, options.basepath)
  File "E:\code\emsdk\emscripten\incoming\tools\wasm-sourcemap.py", line 285, in build_sourcemap
    file_name = os.path.relpath(os.path.abspath(file_name), base_path)
  File "E:\code\emsdk\python\2.7.13.1_64bit\python-2.7.13.amd64\lib\ntpath.py", line 529, in relpath
    % (path_prefix, start_prefix))
ValueError: path is on drive E:, start on drive c:
shared:ERROR: 'E:\code\emsdk\python\2.7.13.1_64bit\python-2.7.13.amd64\python.exe E:\code\emsdk\emscripten\incoming\tools\wasm-sourcemap.py c:\users\jukkaj\appdata\local\temp\emscripten_temp_c0jjbu\test.wasm --dwarfdump=E:/code/emsdk/llvm/git/build_master_vs2019_64/Release/bin\llvm-dwarfdump.exe -o c:\users\jukkaj\appdata\local\temp\emscripten_temp_c0jjbu\test.wasm.map --basepath=c:\users\jukkaj\appdata\local\temp\emscripten_test_browser_sjtjx3' failed (1)
ERROR
[Browser harness server terminated]
```
where
```
file_name = E:/code/emsdk/emscripten/incoming/tests/pthread/test_pthread_lsan_leak.cpp
base_path = c:\users\jukkaj\appdata\local\temp\emscripten_test_browser_sjtjx3
```
hence `os.path.relpath()` throws.

**Comment by juj** [comment]
Can ignore past the error with
```
diff --git a/tools/wasm-sourcemap.py b/tools/wasm-sourcemap.py
index 2f2110887..0cb5f30e3 100755
--- a/tools/wasm-sourcemap.py
+++ b/tools/wasm-sourcemap.py
@@ -279,7 +279,11 @@ def build_sourcemap(entries, code_section_offset, prefixes, collect_sources, bas
     if prefixes.provided():
       source_name = prefixes.sources.resolve(file_name)
     else:
-      file_name = os.path.relpath(os.path.abspath(file_name), base_path)
+      file_name = os.path.abspath(file_name)
+      try:
+        file_name = os.path.relpath(file_name, base_path)
+      except Exception:
+        pass
       file_name = normalize_path(file_name)
       source_name = file_name
     if source_name not in sources_map:
```
but overall the whole logic seems odd - not sure why source maps would be getting generated relative to the temp directory in the first place? (or is this some fluke with the test suite runner only)

**Comment by kripken** [comment]
I think @RReverser can explain best why source map locations must be relative to the wasm file directory (I think that's how the devtools look the files up later).

**Comment by kmcgrath-cd** [comment]
I'm running into this as well with the latest SDK on Windows 10.  The issue isn't that the source map locations must be relative to the wasm file directory (they are in my instance), it's that the emscripten SDK toolchain must be relative to the source your building.  My SDK/toolchain is on my C: drive, but all of my source that I'm building, and the output directory, is on my D: drive, and wasm-sourcemap.py gets the same exception stated originally.

Let me be clearer, my source code that I'm building from is on my D: drive.  My output directory is also on my D: drive.  But my emscripten SDK is on my C: drive, and that's the issue because wasm-sourcemap.py wants to make a relative path from my source code / output drive to my SDK drive, which of course can't work.

I'm trying to find a work around, something I can pass to the emscripten linker, to change the --basepath argument that wasm-sourcemap.py is getting, but I haven't found one yet.

**Comment by kmcgrath-cd** [comment]
Command-line returning the Python exception:
`C:\Python36\python.exe C:\Wasm\emsdk\upstream\emscripten\tools\wasm-sourcemap.py D:\Temp\emscripten_temp_tjodgtfs\porttest.wasm --dwarfdump=C:/Wasm/emsdk/upstream/bin\llvm-dwarfdump.exe -o D:\Temp\emscripten_temp_tjodgtfs\porttest.wasm.map --basepath=D:\Work\PortTest\Code\Wasm\WasmBuild\Debug`

As you can see, the source and even the temporary files are on the D: drive.  The only thing that isn't is the emscripten SDK, which is on the C: drive.  I think it's trying to get a relative path from dwarfdump for some reason I'm failing to understand.

**Comment by RReverser** [comment]
> Let me be clearer, my source code that I'm building from is on my D: drive. My output directory is also on my D: drive. But my emscripten SDK is on my C: drive, and that's the issue because wasm-sourcemap.py wants to make a relative path from my source code / output drive to my SDK drive

Hm yeah that sounds like a bug; there is no reason for wasm-sourcemap.py to care about SDK path.

**Comment by timluther** [comment]
Was going to file this one too. The problem is down to the use of 'os.path.relpath'  which asserts if they're on different drives. I put a try-catch around it and just accepted the fully qualified version of the path if it fails. That seemed to work.

**Comment by kripken** [comment]
Please submit a PR with that if you can @timluther , that sounds like something we should fix in upstream.

**Comment by timluther** [comment]
Hi Kripken,
  My changes are in \upstream\emscripten\tools which doesn't seem to be part of the git repo. Any ideas how to do a pull request on that?

**Comment by kripken** [comment]
@timluther Yes, the `upstream` directory exists in the emsdk folder, but development on the code is done on emscripten (those are separate repos in the `emscripten-core` org). So you can clone the emscripten repo (the one this issue is in), and make a PR with that.

**Comment by clevea** [comment]
This is also causing me problems. I took the patch above and applied it to get past them. In my case emscripten and my source are on D: but emscripten links the wasm int a temp directory on C: which leads to the crash. Initially it was using my users temp directory e.g. `c:\users\cleve\appdata\local\temp...`. I tried setting the environment variables TMP and TEMP to a directory on D:. Then emscripten linked into `c:\temp\...`.

Thanks for the patch.
